pkg/gofr/datasource/surrealdb/surrealdb.go

```diff
@@ -0,0 +1,643 @@
+	err := c.db.Use(c.config.Namespace, c.config.Database)
+	if err != nil {
+		c.logError("unable to set the namespace and database for SurrealDB", err)
+		return err
+	}
+
+	return nil
+}
+
+// SignIn is a helper method for signing in a user.
+func (c *Client) signIn(authData *surrealdb.Auth) (string, error) {
+	var token connection.RPCResponse[string]
+	if err := c.db.Send(&token, "signin", authData); err != nil {
+		return "", err
+	}
+
+	if err := c.db.Let(constants.AuthTokenKey, token.Result); err != nil {
+		return "", err
+	}
+
+	return *token.Result, nil
+}
+
+// authenticate handles the authentication process if credentials are provided.
+func (c *Client) authenticateCredentials() error {
+	if c.config.Username == "" && c.config.Password == "" {
+		return nil
+	}
+
+	if c.config.Username == "" || c.config.Password == "" {
+		return errInvalidCredentialsConfig
+	}
+
+	_, err := c.signIn(&surrealdb.Auth{
+		Username: c.config.Username,
+		Password: c.config.Password,
+	})
+	if err != nil {
+		c.logError("failed to sign in to SurrealDB", err)
+		return err
+	}
+
+	return nil
+}
+
+// logError is a helper function to log errors.
+func (c *Client) logError(message string, err error) {
+	if c.logger == nil {
+		return
+	}
+
+	if err != nil {
+		c.logger.Errorf("%s: %v", message, err)
+		return
+	}
+
+	c.logger.Errorf("%s", message)
+}
+
+// useNamespace switches the active namespace for the database connection.
+func (c *Client) useNamespace(ns string) error {
+	if c.db == nil {
+		return errNotConnected
+	}
+
+	return c.db.Use(ns, "")
+}
+
+// useDatabase switches the active database for the connection.
+func (c *Client) useDatabase(db string) error {
+	if c.db == nil {
+		return errNotConnected
+	}
+
+	return c.db.Use("", db)
+}
+
+type QueryResponse struct {
+	ID     any                  `json:"id" msgpack:"id"`
+	Error  *connection.RPCError `json:"error,omitempty" msgpack:"error,omitempty"`
+	Result *[]QueryResult       `json:"result,omitempty" msgpack:"result,omitempty"`
+}
+
+type QueryResult struct {
+	Status string `json:"status"`
+	Time   string `json:"time"`
+	Result any    `json:"result"`
+}
+
+// Query executes a query on the SurrealDB instance.
+// Query executes a query on the SurrealDB instance.
+func (c *Client) Query(ctx context.Context, query string, vars map[string]any) ([]any, error) {
+	span := c.addTrace(ctx, "Query", query)
+	if span != nil {
+		defer span.End()
+	}
+
+	if c.db == nil {
+		return nil, errNotConnected
+	}
+
+	startTime := time.Now()
+
+	defer c.sendOperationStats(&QueryLog{
+		Query:         query,
+		OperationName: "query",
+		Namespace:     c.config.Namespace,
+		Database:      c.config.Database,
+		Data:          vars,
+		Span:          span,
+	}, startTime)
+
+	var res QueryResponse
+
+	if err := c.db.Send(&res, "query", query, vars); err != nil {
+		return nil, err
+	}
+
+	result := res.Result
+
+	resp := make([]any, 0)
+
+	for _, r := range *result {
+		if r.Status != "OK" {
```

Comment:
This OK could be moved to a constant explaining what is this value, where it comes from.